cmd/ctr/run.go

```diff
@@ -94,82 +99,131 @@ var runCommand = cli.Command{
 		if err != nil {
 			return err
 		}
-
 		imageStore, err := getImageStore(context)
 		if err != nil {
 			return errors.Wrap(err, "failed resolving image store")
 		}
-
-		if runtime.GOOS != "windows" && context.String("rootfs") == "" {
-			ref := context.Args().First()
-
-			image, err := imageStore.Get(ctx, ref)
+		differ, err := getDiffService(context)
+		if err != nil {
+			return err
+		}
+		var (
+			checkpoint      digest.Digest
+			checkpointIndex = digest.Digest(context.String("checkpoint"))
+			ref             = context.Args().First()
+		)
+		var spec []byte
+		if checkpointIndex != "" {
+			var index ocispec.ImageIndex
+			r, err := content.Reader(ctx, checkpointIndex)
 			if err != nil {
-				return errors.Wrapf(err, "could not resolve %q", ref)
+				return err
 			}
-			// let's close out our db and tx so we don't hold the lock whilst running.
-
-			diffIDs, err := image.RootFS(ctx, content)
+			err = json.NewDecoder(r).Decode(&index)
+			r.Close()
 			if err != nil {
 				return err
 			}
-
-			if context.Bool("readonly") {
-				mounts, err = snapshotter.View(ctx, id, identity.ChainID(diffIDs).String())
-			} else {
-				mounts, err = snapshotter.Prepare(ctx, id, identity.ChainID(diffIDs).String())
+			var rw ocispec.Descriptor
+			for _, m := range index.Manifests {
+				switch m.MediaType {
+				case images.MediaTypeContainerd1Checkpoint:
+					checkpoint = m.Digest
+				case images.MediaTypeContainerd1CheckpointConfig:
+					if r, err = content.Reader(ctx, m.Digest); err != nil {
+						return err
+					}
+					spec, err = ioutil.ReadAll(r)
+					r.Close()
+					if err != nil {
+						return err
+					}
+				case images.MediaTypeDockerSchema2Manifest:
+					// make sure we have the original image that was used during checkpoint
+					diffIDs, err := images.RootFS(ctx, content, m.Descriptor)
+					if err != nil {
+						return err
+					}
+					if _, err := snapshotter.Prepare(ctx, id, identity.ChainID(diffIDs).String()); err != nil {
+						if !snapshot.IsExist(err) {
+							return err
+						}
+					}
+				case ocispec.MediaTypeImageLayer:
+					rw = m.Descriptor
+				}
 			}
-			if err != nil {
-				if !snapshot.IsExist(err) {
-					return err
+			if mounts, err = snapshotter.Mounts(ctx, id); err != nil {
```

Comment:
This seems to be handling a case where a snapshot for `id` already exists. We are probably not doing the right thing already since id could be a snapshot of an older image. This case also seems wrong since it will just attempt to re-apply on an existing layer. We don't need to change this behavior right now but we need to find a better way of handling this case.